src/lib.rs

```diff
@@ -179,31 +210,66 @@ impl GpuProfiler {
-        (encoder_or_pass.is_pass() && self.enable_pass_timer) || (!encoder_or_pass.is_pass() && self.enable_encoder_timer)
+    /// Creates a new profiler and connects to a running Tracy client.
+    #[cfg(feature = "tracy")]
+    pub fn new_with_tracy_client(
+        settings: GpuProfilerSettings,
+        backend: wgpu::Backend,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+    ) -> Result<Self, CreationError> {
+        let mut profiler = Self::new(settings)?;
+        profiler.tracy_context = Some(tracy::create_tracy_gpu_client(backend, device, queue)?);
+        Ok(profiler)
     }
 
-    /// Starts a new debug/timer scope on a given encoder or rendering/compute pass.
+    /// Changes the settings of an existing profiler.
     ///
-    /// Scopes can be arbitrarily nested.
+    /// This fails if there are open profiling scopes.
     ///
-    /// May create new wgpu query objects (which is why it needs a [`wgpu::Device`] reference).
+    /// If timer scopes are disabled (by setting [GpuProfilerSettings::enable_timer_scopes] to false),
+    /// any timer queries that are in flight will still be processed,
+    /// but unused query sets and buffers will be deallocated during [`Self::process_finished_frame`].
+    pub fn change_settings(&mut self, settings: GpuProfilerSettings) -> Result<(), SettingsError> {
+        if !self.open_scopes.is_empty() {
+            Err(SettingsError::HasOpenScopes)
+        } else {
+            settings.validate()?;
+            if !settings.enable_timer_scopes {
+                self.unused_pools.clear();
+            }
+            self.settings = settings;
+
+            Ok(())
+        }
+    }
+
+    /// Starts a new debug & timer scope on a given encoder or rendering/compute pass if enabled.
+    ///
+    /// If an [`wgpu::CommandEncoder`] is passed but the [`wgpu::Device`]
+    /// does not support [`wgpu::Features::TIMESTAMP_QUERY`], no scope will be opened.
+    /// If an [`wgpu::ComputePass`] or [`wgpu::RenderPass`] is passed but the [`wgpu::Device`]
+    /// does not support [`wgpu::Features::TIMESTAMP_QUERY_INSIDE_PASSES`], no scope will be opened.
     ///
-    /// If encoder/pass timer queries are disabled respectively, no scope will be opened.
+    /// May allocate a new [`wgpu::QuerySet`] and [`wgpu::Buffer`] internally if necessary.
+    /// After the first call, the same [`wgpu::Device`] must be used with all subsequent calls
+    /// (and all passed references to wgpu objects must originate from that device).
     ///
-    /// See also [`wgpu_profiler!`], [`GpuProfiler::end_scope`]
+    /// See also [`GpuProfiler::end_scope`], [`wgpu_profiler!`]
     #[track_caller]
     pub fn begin_scope<Recorder: ProfilerCommandRecorder>(&mut self, label: &str, encoder_or_pass: &mut Recorder, device: &wgpu::Device) {
-        if self.timestamp_write_enabled(encoder_or_pass) {
+        let features = self.device_features.get_or_insert_with(|| device.features());
```

Comment:
Why are device features lazily populated? Is there an advantage to not taking a &Device during initialization?

Similar question applies to the timestamp_period parameter in process_finished_frame, rather than getting this during initialization.

Reply:
I figured the _only_ reason I had the device passed in on initialization is the features so I might have as well expected the features themselves. But then in turn we _have_ to take the device in `begin_scope` and the first `begin_scope` call is in fact also the first one that needs the information. This results in overall less passing of not-strictly-needed stuff.

The `timestamp_period` is in a similar vain. But having read Chrome's implementation thereof there's actually another nasty detail here: Chrome slowly converges on a value there, meaning the first few calls may have different values than later calls. Again, I wanted to limit the "dependencies" of the interface.

Reply:
on the flip side do you think this makes things more annoying to use? 🤔

Reply:
Ah that makes sense.

I don't think this makes it annoying at all. The device part is transparent to user and needing to provide the timestamp period in one spot is not a major task.

> But having read Chrome's implementation thereof there's actually another nasty detail here: Chrome slowly converges on a value there, meaning the first few calls may have different values than later calls.

I imagine if users aren't aware of this they will might just cache the value from the first call... at least that would be my inclination. I feel like wgpu should document this? And it might be useful to include a note on proccess_finished_frame.